checker/src/org/checkerframework/checker/lock/LockAnnotatedTypeFactory.java

```diff
@@ -105,28 +120,366 @@ public LockTransfer createFlowTransferFunction(CFAbstractAnalysis<CFValue, LockS
+     *
+     * @param am AnnotationMirror for a GuardSatisfied annotation.
+     * @return The index on the GuardSatisfied annotation.
+     */
+    // package-private
+    int getGuardSatisfiedIndex(AnnotationMirror am) {
+        return AnnotationUtils.
+                getElementValue(am, "value", Integer.class, true);
+    }
+
+    @Override
+    public Pair<AnnotatedExecutableType, List<AnnotatedTypeMirror>> methodFromUse(
+            ExpressionTree tree, ExecutableElement methodElt,
+            AnnotatedTypeMirror receiverType) {
+        Pair<AnnotatedExecutableType, List<AnnotatedTypeMirror>> mfuPair = super.methodFromUse(tree, methodElt, receiverType);
+
+        if (tree.getKind() == Kind.METHOD_INVOCATION) {
+            // If a method's formal return type is annotated with @GuardSatisfied(index),
+            // look for the first instance of @GuardSatisfied(index) in the method definition's receiver type or
+            // formal parameters, retrieve the corresponding type of the actual parameter / receiver at the call site
+            // (e.g. @GuardedBy("someLock") and replace the return type at the call site with this type.
+
+            MethodInvocationTree methodInvocationTree = (MethodInvocationTree) tree;
+            AnnotatedExecutableType invokedMethod = mfuPair.first;
+
+            if (invokedMethod.getElement().getKind() != ElementKind.CONSTRUCTOR) {
+                AnnotatedTypeMirror methodDefinitionReturn = invokedMethod.getReturnType();
+
+                if (methodDefinitionReturn != null && methodDefinitionReturn.hasAnnotation(GuardSatisfied.class)) {
+                    int returnGuardSatisfiedIndex = getGuardSatisfiedIndex(methodDefinitionReturn);
+
+                    // @GuardSatisfied with no index defaults to index -1. Ignore instances of @GuardSatisfied with no index.
+                    // If a method is defined with a return type of @GuardSatisfied with no index, an error is reported by LockVisitor.visitMethod.
+                    if (returnGuardSatisfiedIndex != -1) {
+
+                        // Find the receiver or first parameter whose @GS index matches that of the return type.
+                        // Ensuring that the type annotations on distinct @GS parameters with the same index match at the call site is handled in LockVisitor.visitMethodInvocation
+
+                        ExecutableElement invokedMethodElement = invokedMethod.getElement();
+                        if (!ElementUtils.isStatic(invokedMethodElement)) {
+                            AnnotatedTypeMirror methodDefinitionReceiver = invokedMethod.getReceiverType();
+                            if (methodDefinitionReceiver != null && methodDefinitionReceiver.hasAnnotation(GuardSatisfied.class)) {
+                                int receiverGuardSatisfiedIndex = getGuardSatisfiedIndex(methodDefinitionReceiver);
+
+                                if (receiverGuardSatisfiedIndex == returnGuardSatisfiedIndex) {
```

Comment:
I find this level of nesting rather hard to follow.
There is some duplication between the handling of the receiver and the other parameters. Maybe that can be extracted into a helper method to reduce duplication and nesting?